### PR TITLE
fix: Resolving problem where indexer requests weren't retrying

### PR DESCRIPTION
**Pull request by robdmoore**
As reported in #164

### CODE REVIEW

File: src/network-client.spec.ts [hunk 1]
```diff
@@ -2,9 +2,31 @@ import { describe, expect, test } from '@jest/globals'
...
+// Don't spam algonode all the time, remove the `skip` to run these manually
+describe.skip('network-clients', () => {
   envResetFixture()
 
+  describe('Retry', () => {
+    test('Retries indexer calls', async () => {
+      const indexer = await algokit.getAlgoIndexerClient(algokit.getAlgoNodeConfig('testnet', 'indexer'))
+
+      await Promise.all(
+        new Array(70).fill(0).map(async (_) => {
```
Comment (ShoGinn) [inline src/network-client.spec.ts:14]: Unfortunately, this won't kick off the 429 unless you run the test more than once. Upping it to 150 like the other is a good fix.

File: src/network-client.spec.ts [hunk 2]
```diff
@@ -2,9 +2,31 @@ import { describe, expect, test } from '@jest/globals'
...
+      const algod = await algokit.getAlgoClient(algokit.getAlgoNodeConfig('testnet', 'algod'))
+
+      await Promise.all(
+        new Array(150).fill(0).map(async (_) => {
+          await algod.accountInformation('XBYLS2E6YI6XXL5BWCAMOA4GTWHXWENZMX5UHXMRNWWUQ7BXCY5WC5TEPA').do()
+        }),
+      )
+    })
+  })
+
```
Comment (ShoGinn) [inline src/network-client.spec.ts:29]: I took the liberty to update the test suite to be more robust.

```
// Don't spam algonode all the time, remove the `skip` to run these manually
describe.skip('Retry', () => {
    /*
    *https://nodely.io/api/*
    The API requests are rate limited per source IP address with request shaping.
    A burst of 90 rps is allowed and all responses are delayed artificially to make sure that (if called sequentially) they will not exceed the quota.
    Queries exceeding the 90rps in burst or sustained 60rps will get HTTP 429 error.
    */
    // Create a mock logger to track the number of retries
    const myLogger = {
      error: jest.fn(),
      warn: jest.fn(),
      info: jest.fn(),
      debug: jest.fn(),
      verbose: jest.fn(),
    }
    algokit.Config.configure({ logger: myLogger })
    afterEach(() => {
      jest.clearAllMocks()
    })
    test('Retries indexer calls', async () => {
      const indexer = await algokit.getAlgoIndexerClient(algokit.getAlgoNodeConfig('testnet', 'indexer'))

      const response = await Promise.all(
        new Array(150).fill(0).map(async (_) => {
          return await indexer.lookupAccountByID('XBYLS2E6YI6XXL5BWCAMOA4GTWHXWENZMX5UHXMRNWWUQ7BXCY5WC5TEPA').do()
        }),
      )
      expect(response.length).toBe(150)
      expect(myLogger.warn).toHaveBeenCalledWith(
        'algosdk request failed 1 times. Retrying in 0ms: URLTokenBaseHTTPError: Network request error. Received status 429 (Too Many Requests)',
      )
    })
    test('Retries algod calls', async () => {
      const algod = await algokit.getAlgoClient(algokit.getAlgoNodeConfig('testnet', 'algod'))

      const response = await Promise.all(
        new Array(150).fill(0).map(async (_) => {
          return await algod.accountInformation('XBYLS2E6YI6XXL5BWCAMOA4GTWHXWENZMX5UHXMRNWWUQ7BXCY5WC5TEPA').do()
        }),
      )
      expect(response.length).toBe(150)
      expect(myLogger.warn).toHaveBeenCalledWith(
        'algosdk request failed 1 times. Retrying in 0ms: URLTokenBaseHTTPError: Network request error. Received status 429 (Too Many Requests)',
      )
    })
  })
  ```

**Review by ShoGinn — CHANGES_REQUESTED** [review]
Thanks for updating this! I added some suggested fixes in the tests but not in the retry logic changes.

File: src/network-client.spec.ts (hunk 2 above)
Comment (ShoGinn) [inline src/network-client.spec.ts:29]:
```suggestion
describe('network-clients', () => {
  envResetFixture()

// Don't spam algonode all the time, remove the `skip` to run these manually
describe.skip('Retry', () => {
    /*
    *https://nodely.io/api/*
    The API requests are rate limited per source IP address with request shaping.
    A burst of 90 rps is allowed and all responses are delayed artificially to make sure that (if called sequentially) they will not exceed the quota.
    Queries exceeding the 90rps in burst or sustained 60rps will get HTTP 429 error.
    */
    // Create a mock logger to track the number of retries
    const myLogger = {
      error: jest.fn(),
      warn: jest.fn(),
      info: jest.fn(),
      debug: jest.fn(),
      verbose: jest.fn(),
    }
    algokit.Config.configure({ logger: myLogger })
    afterEach(() => {
      jest.clearAllMocks()
    })
    test('Retries indexer calls', async () => {
      const indexer = await algokit.getAlgoIndexerClient(algokit.getAlgoNodeConfig('testnet', 'indexer'))

      const response = await Promise.all(
        new Array(150).fill(0).map(async (_) => {
          return await indexer.lookupAccountByID('XBYLS2E6YI6XXL5BWCAMOA4GTWHXWENZMX5UHXMRNWWUQ7BXCY5WC5TEPA').do()
        }),
      )
      expect(response.length).toBe(150)
      expect(myLogger.warn).toHaveBeenCalledWith(
        'algosdk request failed 1 times. Retrying in 0ms: URLTokenBaseHTTPError: Network request error. Received status 429 (Too Many Requests)',
      )
    })
    test('Retries algod calls', async () => {
      const algod = await algokit.getAlgoClient(algokit.getAlgoNodeConfig('testnet', 'algod'))

      const response = await Promise.all(
        new Array(150).fill(0).map(async (_) => {
          return await algod.accountInformation('XBYLS2E6YI6XXL5BWCAMOA4GTWHXWENZMX5UHXMRNWWUQ7BXCY5WC5TEPA').do()
        }),
      )
      expect(response.length).toBe(150)
      expect(myLogger.warn).toHaveBeenCalledWith(
        'algosdk request failed 1 times. Retrying in 0ms: URLTokenBaseHTTPError: Network request error. Received status 429 (Too Many Requests)',
      )
    })
  })

```

**Review by ShoGinn — CHANGES_REQUESTED** [review]
Forgot about the suggestions feature ;)

File: src/network-client.spec.ts [hunk 3]
```diff
@@ -2,9 +2,31 @@ import { describe, expect, test } from '@jest/globals'
 import { envResetFixture } from '../tests/fixtures/env-fixture'
 import * as algokit from './'
 
-describe('network-clients', () => {
+// Don't spam algonode all the time, remove the `skip` to run these manually
+describe.skip('network-clients', () => {
```
[the remark this answers is not in the document]
Comment (robdmoore) [reply src/network-client.spec.ts:6]: Oops, that's my bad I was moving it around and it ended up in the wrong spot, good pickup!

File: src/network-client.spec.ts (hunk 2 above)
Comment (robdmoore) [reply src/network-client.spec.ts]: Awesome addition thanks!

**Pull request merged**